hive/agents/ppo.py

```diff
@@ -0,0 +1,307 @@
+        super().train()
+        self._actor_critic.train()
+
+    def eval(self):
+        """Changes the agent to evaluation mode."""
+        super().eval()
+        self._actor_critic.eval()
+
+    def preprocess_update_info(self, update_info):
+        """Preprocesses the :obj:`update_info` before it goes into the replay buffer.
+        Clips the reward in update_info.
+
+        Args:
+            update_info: Contains the information from the current timestep that the
+                agent should use to update itself.
+        """
+        preprocessed_update_info = {
+            "observation": update_info["observation"],
+            "action": update_info["action"],
+            "reward": update_info["reward"],
+            "done": update_info["done"],
+        }
+        if "agent_id" in update_info:
+            preprocessed_update_info["agent_id"] = int(update_info["agent_id"])
+
+        return preprocessed_update_info
+
+    def preprocess_update_batch(self, batch):
+        """Preprocess the batch sampled from the replay buffer.
+
+        Args:
+            batch: Batch sampled from the replay buffer for the current update.
+
+        Returns:
+                - Preprocessed batch.
+        """
+        for key in batch:
+            batch[key] = torch.tensor(batch[key], device=self._device)
+        return batch
+
+    @torch.no_grad()
+    def act(self, observation):
+        """Returns the action for the agent. 
+
+        Args:
+            observation: The current observation.
+        """
+
+        observation = torch.tensor(
+            np.expand_dims(observation, axis=0), device=self._device
+        ).float()
+        action, logprob, _, value = self.self._actor_critic.get_action_and_value()
+        
+        if (
+            self._training
+            and self._logger.should_log(self._timescale)
+            and self._state["episode_start"]
+        ):
+            self._logger.log_scalar("train_qval", value[action], self._timescale)
+            self._state["episode_start"] = False
+        
+        
+        self._logprob = logprob
+        self._cur_value = value
+        return action
+
+    def update(self, update_info):
+        """
+        Updates the PPO agent.
+
+        Args:
+            update_info: dictionary containing all the necessary information to
+                update the agent. Should contain a full transition, with keys for
+                "observation", "action", "reward", and "done".
+        """
+        
+        if update_info["done"]:
+            self._state["episode_start"] = True
+
+        if not self._training:
+            return
+        
+        # Inspired  by https://github.com/vwxyzjn/ppo-implementation-details/blob/main/ppo_shared.py
+        # Check if the buffer is already full, if it is use current observation to bootstrap value estimate
+        if (self._replay_buffer.size() >= self._transitions_per_update): #We need next observation
```

Comment:
For PPO I feel we can keep replay_buffer and transitions_per_update the same. We would need something like batch_iterations to reuse the entire replay buffer and then discard it. We don't need a circular history right?

Taking inspiration from https://github.com/fabiopardo/tonic/blob/master/tonic/torch/agents/ppo.py

@kshitijkg @dapatil211

Reply:
We want to perform minibatch gradient descent. So we want to use a minibatch of size batch_size for each update. 

We don't really need a circular history. But I have used that because that's the main replay buffer being used for other algorithms in RLHive. 

Following this blog: https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/